ratpack-zipkin/src/main/java/ratpack/zipkin/ServerTracingModule.java

```diff
@@ -18,118 +18,105 @@
 import ratpack.guice.ConfigurableModule;
-import ratpack.http.Request;
-import ratpack.http.Response;
+import ratpack.handling.HandlerDecorator;
+import ratpack.zipkin.internal.DefaultServerTracingHandler;
 import ratpack.zipkin.internal.ServerRequestAdapterFactory;
 import ratpack.zipkin.internal.ServerResponseAdapterFactory;
 
-import java.util.Collection;
-import java.util.Collections;
+import static com.google.inject.Scopes.SINGLETON;
 
 /**
  * Module for ZipKin distributed tracing.
  */
 public class ServerTracingModule extends ConfigurableModule<ServerTracingModule.Config> {
   @Override
   protected void configure() {
-    bind(ServerTracingHandler.class);
-    bind(ServerTracingDecorator.class);
-    bind(ServerRequestAdapterFactory.class);
-    bind(ServerResponseAdapterFactory.class);
-  }
+    bind(ServerRequestAdapterFactory.class).in(SINGLETON);
+    bind(ServerResponseAdapterFactory.class).in(SINGLETON);
 
-  @Provides
-  public Brave getBrave(final Config config) {
-    return config.brave;
+    bind(ServerTracingHandler.class).to(DefaultServerTracingHandler.class);
+    Provider<ServerTracingHandler> serverTracingHandlerProviderProvider = getProvider(ServerTracingHandler.class);
+    Multibinder.newSetBinder(binder(), HandlerDecorator.class).addBinding().toProvider(() -> HandlerDecorator.prepend(serverTracingHandlerProviderProvider.get()));
   }
 
   @Provides
-  public ServerTracer getServerTracer(final Brave brave) {
-    return brave.serverTracer();
+  public SpanNameProvider spanNameProvider(final Config config) {
+    return config.spanNameProvider;
   }
 
   @Provides
-  public LocalTracer localTracer(final Brave brave) {
-    return brave.localTracer();
+  public RequestAnnotationExtractor requestAnnotationExtractorFunc(final Config config) {
+    return config.requestAnnotationFunc;
   }
 
   @Provides
-  public ServerResponseInterceptor serverResponseInterceptor(final ServerTracer tracer) {
-    return new ServerResponseInterceptor(tracer);
+  public ResponseAnnotationExtractor responseAnnotationExtractorFunc(final Config config) {
+    return config.responseAnnotationFunc;
   }
 
   @Provides
-  public ServerRequestInterceptor serverRequestInterceptor(final ServerTracer tracer) {
-    return new ServerRequestInterceptor(tracer);
+  public ServerResponseInterceptor serverResponseInterceptor(final Brave brave) {
+    return new ServerResponseInterceptor(brave.serverTracer());
   }
 
   @Provides
-  public SpanNameProvider spanNameProvider(final Config config) {
-    return config.spanNameProvider;
+  public ServerRequestInterceptor serverRequestInterceptor(final Brave brave) {
+    return new ServerRequestInterceptor(brave.serverTracer());
   }
 
   @Provides
-  public RequestAnnotationExtractor requestAnnotationExtractorFunc(final Config config) {
-    return config.requestAnnotationFunc;
+  public Brave getBrave(final Config config) {
+    Brave.Builder braveBuilder = new Brave.Builder(config.serviceName);
```

Comment:
The Hystrix module is a good example.  See https://github.com/ratpack/ratpack/blob/master/ratpack-hystrix/src/main/java/ratpack/hystrix/internal/HystrixRegistryBackedRequestVariable.java#L36 and https://github.com/ratpack/ratpack/blob/master/ratpack-hystrix/src/main/java/ratpack/hystrix/internal/HystrixRegistryBackedRequestVariable.java#L39

See how much sense you can make of it, otherwise I'm happy to start something off or explain in more detail.